src/foundation/src/PDFsharp/src/PdfSharp/Pdf.Signatures/PdfSigner.cs

```diff
@@ -0,0 +1,257 @@
+            var contentStart = 0L;
+            var contentEnd = 0L;
+            var rangeStart = 0L;
+            var rangeEnd = 0L;
+            var extraSpace = 0;
+            signatureValue.SignatureContentsWritten = (sigValue, start, end) =>
+            {
+                contentStart = start;
+                contentEnd = end;
+            };
+            signatureValue.SignatureRangeWritten = (sigValue, start, end) =>
+            {
+                rangeStart = start;
+                rangeEnd = end;
+            };
+            document.AfterSave = (writer) =>
+            {
+                extraSpace = writer.Layout == PdfWriterLayout.Verbose ? 1 : 0;
+            };
+            var ms = new MemoryStream();
+            // copy original document to output-stream
+            inputStream.Seek(0, SeekOrigin.Begin);
+            inputStream.CopyTo(ms);
+            // append incremental update
+            document.Save(ms);
+
+            finalDocumentLength = ms.Length;
+
+            contentStart += extraSpace;
+            rangeStart += extraSpace;
+
+            // write new ByteRange array
+            var rangeArrayValue = string.Format(CultureInfo.InvariantCulture, "[0 {0} {1} {2}]",
+                contentStart, contentEnd, finalDocumentLength - contentEnd);
+            Debug.Assert(rangeArrayValue.Length <= rangeEnd - rangeStart);
+            rangeArrayValue = rangeArrayValue.PadRight((int)(rangeEnd - rangeStart), ' ');
+            ms.Seek(rangeStart, SeekOrigin.Begin);
+            var writeBytes = PdfEncoders.RawEncoding.GetBytes(rangeArrayValue);
+            ms.Write(writeBytes, 0, writeBytes.Length);
+
+            // concat the ranges before and after the content-string
+            var lengthToSign = contentStart + finalDocumentLength - contentEnd;
+            var toSign = new byte[lengthToSign];
+            ms.Seek(0, SeekOrigin.Begin);
+            ms.Read(toSign, 0, (int)contentStart);
+            ms.Seek(contentEnd, SeekOrigin.Begin);
+            ms.Read(toSign, (int)contentStart, (int)(finalDocumentLength - contentEnd));
+
+            // do the signing
+            var signatureData = signer.GetSignedCms(toSign, document);
+
+            // move past the '<'
+            ms.Seek(contentStart + 1, SeekOrigin.Begin);
+            // convert signature to string
+            var signHexString = PdfEncoders.ToHexStringLiteral(signatureData, false, false, null);
+            writeBytes = new byte[signHexString.Length - 2];
+            // exclude '<' and '>' from hex-string and overwrite fake value
+            PdfEncoders.RawEncoding.GetBytes(signHexString, 1, signHexString.Length - 2, writeBytes, 0);
+            ms.Write(writeBytes, 0, writeBytes.Length);
+
+            ms.Position = 0;
+
+            document.Dispose();
+
+            return ms;
+        }
+
+        private int GetContentLength()
+        {
+            return signer.GetSignedCms(new MemoryStream(new byte[] { 0 }), document).Length + 10;
```

Comment:
This method causes the encryption to be called twice, I think that is not good solution, especially if the encryption requires external actions then those are called twice...

Reply:
in lopdf this was handled by creating placeholder values which would be changed later: https://github.com/ralpha/pdf_signing/blob/master/src/signature_info.rs#L117-L124

Reply:
The content length is used to reserve space in the PDF, where the _real_ signature is stored in a later step.
The current approach was used to get a fairly safe estimate of the length, without the risk of reserving _too few_ or _too much_ space.
You could certainly skip this step, but then you have to reserve space for the **largest possible** content length of **any** signer that may be used.
Another approach may be to let the user provide a value (e.g. an `EstimatedContentLength` - property in the signature-options)

Reply:
I at least recommend putting it behind an interface so it can be overriden in the client application. In my use case the digital signature is entered by external card reader and the user needs to enter PIN to the card reader software, asking the PIN twice would be real bad, much worse than causing the pdf size to grow a bit more than necessary

Reply:
@Havunen  Yeah, that's one use case i didn't had in mind when creating the PR, my bad.
Would be nice if we would know more about _how_ users might apply signatures; when knowing the use cases, we could design better interface to handle these...